plugins/state_history_plugin/state_history_plugin.cpp

```diff
@@ -95,12 +96,61 @@ struct state_history_plugin_impl : std::enable_shared_from_this<state_history_pl
       std::shared_ptr<state_history_plugin_impl> plugin;
       bool                                       sending  = false;
       bool                                       sent_abi = false;
-      std::vector<std::vector<char>>             send_queue;
       bool                                       need_to_send_update = false;
 
+      std::thread                                                              thr;
+      std::atomic<bool>                                                        send_thread_has_exception = false;
+      std::exception_ptr                                                       eptr;
+      boost::asio::io_context                                                  ctx;
+      boost::asio::io_context::strand                                          work_strand{ctx};
+      boost::asio::executor_work_guard<boost::asio::io_context::executor_type> work_guard =
+          boost::asio::make_work_guard(ctx);
+
+      uint32_t                                       to_send_block_num = 0;
+      std::optional<std::vector<block_position>::const_iterator> position_it;
+
       session(std::shared_ptr<state_history_plugin_impl> plugin)
           : plugin(std::move(plugin)) {}
 
+      void send_thread_fun() {
+         fc_ilog(_log, "send thread started");
+         try {
+            ctx.run();
+         }
+         catch(std::system_error& ex) {
+            if (ex.code() != std::errc::operation_canceled) {
+               fc_elog(_log, "exception thrown from send thread");
+               eptr = std::current_exception();
+               send_thread_has_exception = true;
+            }
+         }
+         catch (...) {
+            fc_elog(_log, "exception thrown from send thread");
+            eptr = std::current_exception();
+            send_thread_has_exception = true;
+         }
+
+         // avoid ~session() to be executed in this thread.
+         app().post( priority::high, [self = this->derived_session().shared_from_this()]() {
```

Comment:
Does this need to be `high` priority? If you are just pushing it to a diff thread, then `low` should be fine, right?

Reply:
That was my first try; however, it didn't work under heavy load.

Reply:
`medium` would be better as it would then be higher than transaction processing.